### PR TITLE
feat: light-prover.js refactoring, support circom in the cli

### PR DESCRIPTION
**Pull request by sergeytimoshin**
1. Prover refactored, light-zk.js/src/transaction/prover.ts moved to its own package at light-protocol/light-prover.js. 

2. New cli commands: 
 – `light circom:init` initialises the Anchor project with simple circom circuits.
  – `light circom:build`builds a circom circuit, generates a proving key and builds the anchor project.
  – `light circom:test` run a test that creates an off-chain proof using prover.js, sends it on-chain and verifies it in the solana program using groth16-solana verifier.

3. Added new test-circo tests to the cli package, tests invoke previous commands in order.

4. Created new circom-anchor-template repo: https://github.com/Lightprotocol/circom-anchor-template/. 
5. We need to update prover.js dependency in package.json inside of that repo: https://github.com/Lightprotocol/circom-anchor-template/blob/main/package.json#L14  after we publish prover to npmjs registry.

### CODE REVIEW

File: cli/src/psp-utils/buildCircom.ts [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+/**
+ * Generates a zk-SNARK circuit given a circuit name.
+ * Downloads the required powers of tau file if not available.
+ * Compiles the circuit, performs the groth16 setup, and exports the verification key.
+ * Cleans up temporary files upon completion.
+ * @param circuitName - The name of the circuit to be generated.
+ * @returns {Promise<void>}
+ */
+async function generateCircuit({
```
Comment (ananas-block) [inline cli/src/psp-utils/buildCircom.ts:18]: can we unify this function different with generateCircuit in buildPsp?

File: light-prover.js/tests/prover.test.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+const chai = require("chai");
+const chaiAsPromised = require("chai-as-promised");
+chai.use(chaiAsPromised);
+
+describe("Prover Functionality Tests", () => {
+  it("Valid proof test", async () => {
+    const poseidon = await circomlibjs.buildPoseidon();
+    const hash = poseidon.F.toString(poseidon([123]));
+    console.log(hash);
```
Comment (ananas-block) [inline light-prover.js/tests/prover.test.ts:14]: leftover log

File: light-prover.js/tests/prover.test.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    const proofInputs: any = {
+      x: 123,
+      hash: hash,
+    };
+
+    const prover = new Prover(IDL, circuitsPath);
+
+    await prover.addProofInputs(proofInputs);
+    await prover.fullProve();
+    await prover.getVkey();
```
Comment (ananas-block) [inline light-prover.js/tests/prover.test.ts:26]: is this necessary?

File: light-prover.js/tests/prover.test.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    console.time("Proof generation + Parsing");
+    await prover.fullProveAndParse();
+    console.timeEnd("Proof generation + Parsing");
+    // @ts-ignore
+    globalThis.curve_bn128.terminate();
```
Comment (ananas-block) [inline light-prover.js/tests/prover.test.ts:32]: what do you think about making this part of the prover class?

File: light-zk.js/tests/prover.test.ts [hunk 5]
```diff
@@ -1,67 +1,59 @@
 import { assert, expect } from "chai";
-import { Prover } from "../src/transaction/prover";
-import { Idl } from "@coral-xyz/anchor";
-
+import { it } from "mocha";
+import { Prover } from "@lightprotocol/prover.js";
+import * as anchor from "@coral-xyz/anchor";
 let circomlibjs = require("circomlibjs");
 import { Keypair as SolanaKeypair } from "@solana/web3.js";
-import { BN } from "@coral-xyz/anchor";
```
Comment (ananas-block) [inline light-zk.js/tests/prover.test.ts:7]: Since Matteos cleanup we are trying to stick to BN instead of anchor.BN.
Please stick to BN :)

**Review by ananas-block — CHANGES_REQUESTED** [review]

File: cli/src/psp-utils/buildCircom.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+  const match = input.match(regex);
+
+  return match ? match[1] : null;
+}
+
+/**
+ * Searches for a file with the ".light" extension in a specified directory.
+ * Throws an error if more than one such file or no such file is found.
+ * @param directory - The directory to search for the .light file.
+ * @returns {string} - The name of the .light file found in the directory.
```
[the remark this answers is not in the document]
Comment (sergeytimoshin) [reply cli/src/psp-utils/buildCircom.ts]: > A quick fix might be to allow users to pass in the path of the circom file they want to compile and to skip findCircomFile in that case.

Yes, agree, I'll do it in another PR.

File: light-prover.js/tests/prover.test.ts (hunk 4 above)
Comment (sergeytimoshin) [reply light-prover.js/tests/prover.test.ts]: Agreed, moved to prover.

File: light-prover.js/tests/prover.test.ts (hunk 3 above)
Comment (sergeytimoshin) [reply light-prover.js/tests/prover.test.ts]: No, cleaned it up.

File: light-prover.js/tests/prover.test.ts (hunk 2 above)
Comment (sergeytimoshin) [reply light-prover.js/tests/prover.test.ts]: Removed.

File: cli/src/psp-utils/buildCircom.ts (hunk 1 above)
Comment (sergeytimoshin) [reply cli/src/psp-utils/buildCircom.ts]: Yes, removed it from buildPsp, this is circom stuff, so I left it here.
buildPsp.ts file shrinked to one function: https://github.com/Lightprotocol/light-protocol/blob/sergey/prover/cli/src/psp-utils/buildPSP.ts

File: cli/src/psp-utils/buildCircom.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  console.log("🛠  Building on-chain program");
+  await executeAnchor({ args: ["build"] });
+  console.log("✅ Build finished successfully");
+}
+
+export function toSnakeCase(str: string): string {
+  return str.replace(/-/g, "_");
+}
+
+export const snakeCaseToCamelCase = (
```
[the remark this answers is not in the document]
Comment (sergeytimoshin) [reply cli/src/psp-utils/buildCircom.ts]: Moved to [psp-utils/utils](https://github.com/Lightprotocol/light-protocol/blob/sergey/prover/cli/src/psp-utils/utils.ts).

**Review by SwenSchaeferjohann — APPROVED** [review]

**Review by ananas-block — APPROVED** [review]

**Review by vadorovsky — APPROVED** [review]

**Pull request merged**